### PR TITLE
Integrate with Kafka Roller

### PR DESCRIPTION
**Pull request by tinaselenge**
### Type of change

_Select the type of your PR_

- Enhancement / new feature

### Description

* Fix existing altering config test
* Add desired config to TestingKafkaRoller
* Fix KafkaAgentClient call to node config endpoint
* Roll all unready pods, both broker and controllers before rolling the ready pods.
* Skip the admin "isController" check if the node is KRaft controller
* Fix Controller diff check when first changed config is not a controller config
* Add logs
* Add quorum check to be done before rolling controllers
* Update controller config list

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Write tests
- [x] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        if (leaderLastCaughtUpTimestamp < 0) {
+            LOGGER.warnCr(reconciliation, "No valid lastCaughtUpTimestamp is found for the leader replica {} ", leaderId);
+            return false;
+        }
+        LOGGER.debugCr(reconciliation, "The lastCaughtUpTimestamp for the leader replica {} is {}", leaderId, leaderLastCaughtUpTimestamp);
+        replicaStates.forEach((replicaId, lastCaughtUpTimestamp) -> {
+            if (lastCaughtUpTimestamp < 0) {
+                LOGGER.warnCr(reconciliation, "No valid lastCaughtUpTimestamp is found for the replica {} ", replicaId);
+            } else {
+                if (leaderLastCaughtUpTimestamp - lastCaughtUpTimestamp < Long.parseLong(controllerQuorumFetchTimeoutMs)) {
```
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java:73]: We are parsing the same string multiple times here.
I would declare `controllerQuorumFetchTimeoutMs` as a Long and parsing the received String in the constructor just one time (instead of assigning the String) or I would change the constructor to get a Long and parsing the String value in the KafkaRoller code (where the `CONTROLLER_QUORUM_FETCH_TIMEOUT_MS_CONFIG_NAME` is read and assigned to the restartContext).

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java [hunk 2]
```diff
@@ -262,6 +267,7 @@ protected static class RestartContext {
         boolean forceRestart;
         KafkaBrokerConfigurationDiff diff;
         KafkaBrokerLoggingConfigurationDiff logDiff;
+        String controllerQuorumFetchTimeoutMs = "1000";
```
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java:270]: let's define a constant for this "1000" ms default value.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    KafkaQuorumCheck(Reconciliation reconciliation, Admin ac) {
+        this.ac = ac;
+        this.reconciliation = reconciliation;
+        this.quorumInfoFuture = describeMetadataQuorum();
+    }
+
+    /**
+     * Determine whether the given controller can be rolled without affecting the quorum
+     */
+    Future<Boolean> canRoll(int podId) {
```
[the remark this answers is not in the document]
Comment (tinaselenge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java:38]: No specific reason here, other than following the existing pattern of having a package scoped method that calls a private method. We can make canRollController() accessible and directly call it instead of canRoll().

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        LOGGER.debugCr(reconciliation, "Determining whether controller {} can be rolled", podId);
+        return canRollController(quorumInfoFuture, podId);
+    }
+
+    private Future<Boolean> canRollController(Future<QuorumInfo> quorumInfo, int podId) {
```
[the remark this answers is not in the document]
Comment (tinaselenge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java:43]: You are right, we set it in the constructor so we don't need to pass it to the method.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        }).recover(error -> {
+            LOGGER.warnCr(reconciliation, "Error determining whether it is safe to restart pod {}", podId, error);
+            return Future.failedFuture(error);
+        });
+    }
+
+    private boolean wouldAffectQuorum(int leaderId, Map<Integer, Long> replicaStates) {
+        int totalNumVoters = replicaStates.size();
+        AtomicInteger numOfCaughtUpVoters = new AtomicInteger();
+        long controllerQuorumFetchTimeoutMs = 1000;
```
[the remark this answers is not in the document]
Comment (tinaselenge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java]: That was the intention. It's fixed now.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        return canRollController(quorumInfoFuture, podId);
+    }
+
+    private Future<Boolean> canRollController(Future<QuorumInfo> quorumInfo, int podId) {
+        return quorumInfo.map(info -> {
+            int leaderId = info.leaderId();
+            Map<Integer, Long> replicaIdToLastCaughtUpTimestampMap = info.voters().stream().collect(Collectors.toMap(
+                    QuorumInfo.ReplicaState::replicaId,
+                    replicaState -> replicaState.lastCaughtUpTimestamp().isPresent() ? replicaState.lastCaughtUpTimestamp().getAsLong() : -1));
+            boolean canRoll = wouldAffectQuorum(leaderId, replicaIdToLastCaughtUpTimestampMap);
```
[the remark this answers is not in the document]
Comment (tinaselenge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java:49]: You are right, the naming is confusing! How about noImpactOnQuorum or quorumHealthy?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java (hunk 1 above)
Comment (tinaselenge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java:73]: good point!

**Comment by tinaselenge** [comment]
Thank you @ppatierno for reviewing the PR. I think I addressed them, please let me know if there is any further comments.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java]: Yeah `quorumHealthy` matches much better with `canRoll` assignment. I am fine with it.
PS: `isQuorumHealthy` could be the one ... but no need to change again ;-)

**Comment by tombentley** [comment]
@tinaselenge is there some background info/rationale for these changes written down somewhere? Could you add it to the PR description?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java [hunk 7]
```diff
@@ -218,18 +220,21 @@ public Future<Void> rollingRestart(Function<Pod, RestartReasons> podNeedsRestart
...
             for (NodeRef node : nodes)  {
                 // Order the nodes unready first otherwise repeated reconciliations might each restart a pod
                 // only for it not to become ready and thus drive the cluster to a worse state.
-
-                // TODO: In KRaft mode, We currently roll only nodes with the broker role. This is because of Kafka
-                //       limitations. Once managing controller nodes is supported with Kafka Admin API, this should be
-                //       fixed. This is tracked in https://github.com/strimzi/strimzi-kafka-operator/issues/8593.
-                if (node.broker()) {
-                    pods.add(podOperations.isReady(namespace, node.podName()) ? pods.size() : 0, node);
+                // in KRaft mode Roll controllers first, then brokers, but roll unready brokers before ready controllers
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java:228]: "but roll" makes this sentence a little hard to parse. Better to just document the order as a list. 

I think there's a problem with rolling unhealthy brokers first: If the controller quorum is also not healthy then a restarted broker might not be able to register itself with the cluster (because doing so requires a metadata write which is impossible because the controller quorum is not working). 

Of the top of my head I'd suggest the following ordering:

1. Unhealthy controllers (active or not).
2. Healthy non-active controllers that require a restart.
3. Active controller that requires a restart (unless already restarted).
4. Unhealthy pure brokers.
5. Healthy pure brokers that require a restart.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java [hunk 8]
```diff
@@ -375,13 +381,15 @@ private void restartIfNecessary(NodeRef nodeRef, RestartContext restartContext)
...
                 await(isReady(pod), operationTimeoutMs, TimeUnit.MILLISECONDS, e -> new RuntimeException(e));
             } catch (Exception e) {
-                //Initialise the client for KafkaAgent if pod is not ready
-                if (kafkaAgentClient == null) {
-                    this.kafkaAgentClient = initKafkaAgentClient();
-                }
-                BrokerState brokerState = kafkaAgentClient.getBrokerState(pod.getMetadata().getName());
-                if (brokerState.isBrokerInRecovery()) {
-                    throw new UnforceableProblem("Pod " + nodeRef.podName() + " is not ready because the broker is performing log recovery. There are  " + brokerState.remainingLogsToRecover() + " logs and " + brokerState.remainingSegmentsToRecover() + " segments left to recover.", e.getCause());
+                if (nodeRef.broker()) {
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java:384]: Do we not give pure controllers a chance to recover?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java [hunk 9]
```diff
@@ -527,53 +533,78 @@ private void checkReconfigurability(NodeRef nodeRef, Pod pod, RestartContext res
...
-            if (restartContext.backOff.done()) {
+        if (nodeRef.controller()) {
+            LOGGER.debugCr(reconciliation, "Pod {} is a KRaft controller, checking if it requires a restart.", nodeRef);
+            if (kafkaAgentClient == null) {
+                kafkaAgentClient = initKafkaAgentClient();
+            }
+            Config controllerConfig = kafkaAgentClient.getNodeConfiguration(nodeRef.podName());
+            // This configuration property is used by the quorum check for rolling KRaft controllers
+            // and users are allowed to configure this property therefore we need to retrieve its current value.
+            Optional.ofNullable(controllerConfig.get(CONTROLLER_QUORUM_FETCH_TIMEOUT_MS_CONFIG_NAME)).ifPresent(config -> restartContext.controllerQuorumFetchTimeoutMs = Long.getLong(config.value()));
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java:544]: Shouldn't it be the agent's responsibility to supply the default?

Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java:544]: If you are referring to the Kafka agent, how can it know the defaults? I mean, the agent is just returning the controller configuration from a file on the node, it doesn't have any knowledge of defaults for not specified values.

But thinking more about your point ... if we see the Kafka Agent as a "replacement" for the Kafka Admin (for this temporary workaround), maybe we should "hardcode" the defaults (got from documentation/Kafka codebase) in the agent itself? Should the agent return the content of the configuration file on the node + all other missing parameters (with corresponding defaults as values) as the roller was calling the Kafka Admin client?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        });
+    }
+
+    private boolean isQuorumHealthy(int leaderId, Map<Integer, Long> replicaStates) {
+        int totalNumVoters = replicaStates.size();
+        AtomicInteger numOfCaughtUpVoters = new AtomicInteger();
+        long leaderLastCaughtUpTimestamp = replicaStates.get(leaderId);
+        if (leaderLastCaughtUpTimestamp < 0) {
+            LOGGER.warnCr(reconciliation, "No valid lastCaughtUpTimestamp is found for the leader replica {} ", leaderId);
+            return false;
```
[the remark this answers is not in the document]
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java:61]: I see Kafka Java documentation stating this ...

> 
public [OptionalLong](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/OptionalLong.html) lastCaughtUpTimestamp()
Return the last millisecond timestamp at which this replica was known to be caught up with the leader.
Returns:
The value of the lastCaughtUpTime if known, empty otherwise

It seems that returning "empty" is possible, not sure when it could happen.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+    protected Future<QuorumInfo> describeMetadataQuorum() {
+        Promise<QuorumInfo> descPromise = Promise.promise();
+        ac.describeMetadataQuorum().quorumInfo().whenComplete(
+                ((quorumInfo, error) -> {
+                    if (error != null) {
+                        descPromise.fail(error);
+                    } else {
+                        descPromise.complete(quorumInfo);
+                    }
```
[the remark this answers is not in the document]
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java:89]: Good catch Tom, @tinaselenge yes we could use `VertxUtil.kafkaFutureToVertxFuture` method.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (tinaselenge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java:61]: I think this is possible, if the leader has just changed and follower has not started fetching from the new leader yet, either for a very short time or if there is a problem with the follower.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java (hunk 7 above)
Comment (tinaselenge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java:228]: We went back and forth with the order, initially we were doing what you have suggested (unready controllers, ready controllers, unready brokers and ready brokers), which I think makes sense and happy to go with. 

The reason we changed the order was because we thought if we are rolling unready controllers first, that should bring the quorum to healthy state. If there are unready brokers, rolling controllers could bring the cluster to worse state so perhaps, rolling everything that are unready first to get the cluster to healthy state before progressing with any of the ready pods. 

One thing, we are missing in this PR is deferring the active controller so that we roll it last from controller list or full list of pods. At the beginning we were not sure, what's the best way to identify the active controller but now that we have implemented describeMetadataQuorum call to the Admin API, we can do that easily.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java (hunk 9 above)
Comment (tinaselenge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java]: +1 to what @ppatierno said. Currently we have no way of knowing the default values from KafkaAgent. In KafkaRoller, we could make describeConfigs() call to normal brokers' admin API (we already do that for quorum check), to figure out the defaults but that would be an additional call to the brokers.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import org.apache.kafka.clients.admin.Admin;
+import org.apache.kafka.clients.admin.QuorumInfo;
+
+/**
+ * Determines whether the given controller can be rolled without affecting the quorum.
```
[the remark this answers is not in the document]
Comment (tinaselenge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java:18]: Good point! I will make this change, but have just one thing to clarify. We will be doing this quorum check for combined pods as well, not just the pure controllers.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java:61]: But here we're talking about the leaders own state wrt _itself_. In other words, can the leader ever not be cause up to itself?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java (hunk 9 above)
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java]: OK, let's keep this as it is for now.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java:61]: That's actually true :-)

[the remark this answers is not in the document]
Comment (tinaselenge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java:61]: I guess not. If we have a leaderId, then we probably would always have lastCaughtUpTimestamp so perhaps this check does not add any value.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java (hunk 7 above)
Comment (katheris) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java]: We've changed the order to be as you suggested up above. The upshot of the change is that if we fail to roll the KRaft controllers, we don't try to roll the brokers at all. We've separated them into two different Future compositions.

**Comment by tinaselenge** [comment]
@ppatierno should I close this PR given that the [proposal](https://github.com/strimzi/proposals/pull/82) for this is closed?
